owlkettle/playground.nim

```diff
@@ -123,165 +126,142 @@ method view (state: DateDialogState): Widget =
-          index: The index on the list of entries in `state.<fieldName>`
-          typ: The type for which `toListFormField` shall function.
-          
-          Implementing the proc will override this dummy Widget.
-          See the playground module for examples.
-        """
-
-proc toListFormField(state: auto, fieldName: static string, index: int, typ: typedesc[ScaleMark]): Widget =
-  ## Provides a form to display a single entry of type `ScaleMark` in a list of `ScaleMark` entries.
-  let mark: ScaleMark = state.getField(fieldName)[index]
-  return gui:
-    ActionRow:
-      title = fieldName & $index
       
       FormulaEntry {.addSuffix.}:
-        value = mark.value
+        value = field[].value
         xAlign = 1.0
         maxWidth = 8
         proc changed(value: float) =
-          state.getField(fieldName)[index].value = value
+          field[].value = value
       
       DropDown {.addSuffix.}:
         items = ScalePosition.items.toSeq().mapIt($it)
-        selected = mark.position.int
+        selected = field[].position.int
         proc select(enumIndex: int) =
-          state.getField(fieldName)[index].position = enumIndex.ScalePosition
-          
-      Button {.addSuffix.}:
-        icon = "user-trash-symbolic"
-        proc clicked() =
-          state.getField(fieldName).delete(index)
+          field[].position = enumIndex.ScalePosition
 
-proc toFormField[T](state: auto, fieldName: static string, typ: typedesc[seq[T]]): Widget =
-  ## Provides a form field for any field on `state` with a seq type.
-  ## Displays a dummy widget if there is no `toListFormField` implementation for type T.
+proc toFormField[T](state: Viewable, field: ptr seq[T], fieldName: string): Widget =
+  ## Provides a form field for any seq type
+  let formFields = collect(newSeq):
+    for index, value in field[]:
+      toFormField(state, field[][index].addr, fmt"{fieldName} {index}")
+      
   return gui:
     ExpanderRow:
       title = fieldName
       
-      for index, num in state.getField(fieldName):
-        insert(toListFormField(state, fieldName, index, T)){.addRow.}
+      for index, formField in formFields:
+        Box(orient = OrientX) {.addRow.}:
+          insert(formField)
+          Button() {.expand: false.}:
+            icon = "user-trash-symbolic"
+            style = [ButtonDestructive]
+            proc clicked() =
+              field[].delete(index) 
       
       ListBoxRow {.addRow.}:
         Button:
           icon = "list-add-symbolic"
           style = [ButtonFlat]
           proc clicked() =
-            state.getField(fieldName).add(default(T))
+            field[].add(default(T))
+
+template getIterator*(a: typed): untyped =
+  ## Provides a fieldPairs iterator for both ref-types and value-types
+  when a is ref:
+    a[].fieldPairs
+    
+  else:
+    a.fieldPairs
+    
+proc toFormField(state: Viewable, field: ptr[auto], fieldName: string): Widget =
```

Comment:
This proc errors out for variant types. Instead it should show the dummy field as outlined in the doc comment.

Reply:
Why not simply introduce a Variant concept so that variant-types get their own `toFormField` proc which returns the dummy FormField widget?
I'll cook up an example in a second

Reply:
The most recent comment should've solved it. It introduces an ObjectVariant concept that "catches" all toFormField invocation for Object Variants without their own specific toFormField overloads.